vpr/src/route/route_timing.cpp

```diff
@@ -1960,6 +1960,16 @@ void Router::timing_driven_expand_neighbours(t_heap* current,
     RREdgeId first_edge = rr_nodes_->first_edge(from_node);
     RREdgeId last_edge = rr_nodes_->last_edge(from_node);
     int num_edges = size_t(last_edge) - size_t(first_edge);
+    for (int iconn = 0; iconn < num_edges; iconn++) {
+        RREdgeId from_edge(size_t(first_edge) + iconn);
+
+        RRNodeId to_node = rr_nodes_->edge_sink_node(from_edge);
+        rr_nodes_->prefetch_node(to_node);
+
+        int switch_idx = rr_nodes_->edge_switch(from_edge);
+        __builtin_prefetch(&rr_switch_inf_[switch_idx], 0, 0);
+    }
+
```

Comment:
Is it worth wrapping this in an `#ifdef` since it uses an intrinsic that may not perform the same or be present on all CPU architectures?

Reply:
Good question. Will this intrinsic compile on a non-x86 architeture?  OK if it is ignored, but not good to have a compiler error.  If it won't compile on other platforms, it should be wrapped in an #ifdef.

Reply:
> Will this intrinsic compile on a non-x86 architeture?

Yes

Reply:
Thanks.  Follow-up: will it compile in compilers other than g++ (e.g. if you used IBM's compiler to target Power8, would this line error out)?

Reply:
Both GCC and clang have support for `__builtin_prefetch`, other targets will vary.

Reply:
If we're going to put stuff like this in (which I'm not sure is worth it), I'd like to see it wrapped up in a function (e.g. something like `vtr::prefetch()`). That way we can #ifdef it to a no-op in one location on platforms which don't support it.

Reply:
> which I'm not sure is worth it

What's your claim here?  I have evidence that it is worthwhile, so I'm not sure what your statement means.

Reply:
> I'd like to see it wrapped up in a function (e.g. something like `vtr::prefetch()`).

Done, in a macro (because it has to be as far as I can tell).

Reply:
> What's your claim here? I have evidence that it is worthwhile, so I'm not sure what your statement means.

To me there is a tension here, between code generality/maintainability and run-time performance.

My concern is that including these types of actions (e.g. SW prefetching) are generally pretty fragile, mysterious, and tend not to age well. The motivation/reasoning/measurements of why they were added fade overtime, it becomes unclear what assumptions make it necessary, and is unclear when changes in those assumptions (which could occur far away from this code) would invalidate using it (ie when it should be removed). Another concern is that these types of optimizations may be over-fitting the code base to the specific platform on which they are developed/tested.

Now, if the performance benefits are substantial and the above considerations are well accounted for and documented then it may be acceptable, but it depends.

Reply:
> Now, if the performance benefits are substantial and the above considerations are well accounted for and documented then it may be acceptable, but it depends.

I was showing around a 6-8% wallclock runtime difference on directrf and gsm that was repeatable and consistent.  It also makes sense given the access pattern of the core expansion loop.

Reply:
> My concern is that including these types of actions (e.g. SW prefetching) are generally pretty fragile, mysterious, and tend not to age well.

The ideal case would be to write the inner most expansion loop such that the hardware prefetcher can handle the access pattern.  I tried 3-4 refactoring to achieve similiar performance to the simple prefetch loop, and the prefetch loop outperformed all of them.